presto-main/src/main/java/com/facebook/presto/execution/TaskThresholdMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,207 @@
+            Supplier<List<SqlTask>> allTasksSupplier,
+            Function<TaskId, SqlTask> taskSupplier,
+            ScheduledExecutorService taskManagementExecutor,
+            long maxRevocableMemoryPerTask)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.allTasksSupplier = requireNonNull(allTasksSupplier, "allTasksSupplier is null");
+        this.taskSupplier = requireNonNull(taskSupplier, "taskSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+        this.maxRevocableMemoryPerTask = maxRevocableMemoryPerTask;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerTaskMemoryPeriodicCheck();
+        registerPoolListeners();
+    }
+
+    private void registerTaskMemoryPeriodicCheck()
+    {
+        this.scheduledFuture = taskManagementExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                revokeHighMemoryTasksIfNeeded();
+            }
+            catch (Exception e) {
+                log.error(e, "Error requesting task memory revoking");
+            }
+        }, 1, 1, SECONDS);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+            scheduledFuture = null;
+        }
+
+        memoryPools.forEach(memoryPool -> memoryPool.removeTaskRevocableMemoryListener(taskRevocableMemoryListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addTaskRevocableMemoryListener(taskRevocableMemoryListener));
+    }
+
+    @VisibleForTesting
+    void revokeHighMemoryTasksIfNeeded()
+    {
+        if (checkPending.compareAndSet(false, true)) {
+            revokeHighMemoryTasks();
+        }
+    }
+
+    private void onMemoryReserved(TaskId taskId)
+    {
+        try {
+            SqlTask task = taskSupplier.apply(taskId);
+            if (!memoryRevokingNeeded(task)) {
+                return;
+            }
+
+            if (checkPending.compareAndSet(false, true)) {
+                log.debug("Scheduling check for %s", taskId);
+                scheduleRevoking();
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error when acting on memory pool reservation");
+        }
+    }
+
+    private void scheduleRevoking()
+    {
+        taskManagementExecutor.execute(() -> {
+            try {
+                revokeHighMemoryTasks();
+            }
+            catch (Exception e) {
+                log.error(e, "Error requesting memory revoking");
+            }
+        });
+    }
+
+    private boolean memoryRevokingNeeded(SqlTask task)
+    {
+        return task.getTaskContext().filter(taskContext -> taskContext.getTaskMemoryContext().getRevocableMemory() >= maxRevocableMemoryPerTask).isPresent();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *getTaskMemoryContext* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at line 171 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).